proposals/lip-0039.md

```diff
@@ -302,273 +328,379 @@ proof={siblingHashes:[cc956a85, 3c516152, e041e1c0, 6400721e, 3c32b131],
+        return False
+
+    for key, query in zip(queryKeys, proof.queries):
+        # This is the only difference between an inclusion and a non-inclusion proof.
+        if isInclusionProof(key, query) == False:
+            return False
+
+    return verify(queryKeys, proof, root)
 ```
 
-```java
-// This function checks whether two queries correspond to nodes that are
-// children of the same branch node, with q1 and q2 the left and right
-// child respectively
-function areSiblingQueries(q1, q2):
-    if length of q1.binaryBitmap != length of q2.binaryBitmap:
-        return false
-    let h=length of q1.binaryBitmap
-    let binaryKey1=binaryExpansion(q1.key)
-    let binaryKey2=binaryExpansion(q2.key)
-    let keyPrefix1 be the first (h-1) digits of binaryKey1
-    let keyPrefix2 be the first (h-1) digits of binaryKey2
-    if keyPrefix1 != keyPrefix2:
-        return false
-    let d1 be the digit at position h of binaryKey1
-    let d2 be the digit at position h of binaryKey2
-    return d1==0 and d2==1
+```python
+def smtVerifyNonInclusionProof(queryKeys: list[bytes], proof: Proof, root: bytes) -> bool:
+    # Check that there is a query for each key.
+    if len(queryKeys) != len(proof.queries):
+        return False
+
+    for key, query in zip(queryKeys, proof.queries):
+        # This is the only difference between an inclusion and a non-inclusion proof.
+        if isInclusionProof(key, query) == True:
+            return False
+            
+    return verify(queryKeys, proof, root)
 ```
 
-```java
-function calculateRoot(siblingHashes, queries):
-    // we add the hash property to each query, used to recompute the tree root
-    for each query q in queries:
-        set q.binaryBitmap to binaryExpansion(q.bitmap) without leading zeros
-        // if value is empty, it is a defaultEmptyNode
-        if q.value is empty bytes:
-            set q.hash to EMPTY_HASH
+```python
+def verify(queryKeys: list[bytes], proof: Proof, root: bytes) -> bool:
+    for key, query in zip(queryKeys, proof.queries):
+        # The bitmap does not contain extra leading 0s.
+        if query.bitmap[0] == 0:
+            return False
+        if len(key) != KEY_LENGTH_BYTES:
+            return False
+
+        # q is an inclusion proof for key or a default empty node.
+        if key == query.key:
+            continue
+
+        # q is an inclusion proof for another leaf node.
+        commonPrefix = os.path.commonprefix([binaryExpansion(query.key), binaryExpansion(key)])
+        if len(query.binaryBitmap) > len(commonPrefix):
+            # q does not give an non-inclusion proof for key.
+            return False
+
+    filteredQueries = []
+    _filter: dict[bytes, Query] = {}
+    for q in proof.queries:
+        # Remove duplicate queries preserving order.
+        # This can happen if the same query is given for different queryKeys, as for inclusion proofs or non-inclusion proofs pointing to a leaf node
```

Comment:
We do not allow mixing inclusion and non-inclusion proofs, right?

Reply:
The verify function allows it, the new APIs do not allow it

Reply:
But what `verify` actually work correctly if we mix inclusion and non-inclusion proofs? I thought that was the problem and therefore the new public functions distinguish these cases and verify is an internal/private function that may assumed to be called only for inclusion or non-inclusion proofs.

Reply:
It works correctly if you intend its usage to just calculate a root from a proof and verify that it matches another root. It does not distinguish whether the proof is an inclusion or non-inclusion proof.